pricegraph/src/graph/bellman_ford.rs

```diff
@@ -15,8 +16,22 @@ pub type Paths<G> = (
     Vec<Option<<G as GraphBase>::NodeId>>,
 );
 
-/// A negative cycle error with a path representing the cycle.
-pub struct NegativeCycle<G: Data>(pub Vec<G::NodeId>);
+/// A negative cycle error with the node on which it was detected along with the
+/// predecessor vector that can be used to re-create the cycle path.
+pub struct NegativeCycle<G: Data>(pub Vec<Option<G::NodeId>>, pub G::NodeId);
```

Comment:
Also, can you give some context why this was changed in the first place? I found the solution before much more contained (not exposing any option/expect outside the bellman_ford implementation).

I can see that we want to check if the cycle is connected to the fee token. What would happen if we reduced cycles that are not connected to the fee token?

Reply:
So `Option` is used for predecessors since it could be that nodes aren't connected to the starting node (when there are disconnected subgraphs). The issue with just returning the negative cycle is that this information was lost. Since we are reducing each disconnected subgraph to remove overlapping ring trade, this information is needed.

Reply:
> Would it make sense to filter out all Nones and make this a mapping instead?

Yes, that is what this vector is. The reason it was implemented like this is the `bellman_ford::search` was taken from `petgraph::algo::bellman_ford` implementation without modifications besides returning the predecessor vector on failure (so the negative cycle can be reconstructed). We can change this to use `HashMap` for example, but that should probably be in a separate PR where we re-implement the Bellman-Ford search algorithm.

Reply:
>  Since we are reducing each disconnected subgraph to remove overlapping ring trade, this information is needed.

My thought was that even in the implementation before we were returning the cycles (just not the predecessors). As far as I understood the code, we only need the predecessors to check if the cycle is connected to the fee token for the subgraph. Or am I missing some other use case for the predecessor list as part of the NegativeCycle error?

If so, then my question is: What bad would happen if we reduced the cycles regardless of wether or not they are connected to the fee token?

This would allow us to keep the find_cycle logic private to our version of the bellman_ford search.

Reply:
> Or am I missing some other use case for the predecessor list as part of the NegativeCycle error

It is also used for determining disconnected subgraphs

> What bad would happen if we reduced the cycles regardless of wether or not they are connected to the fee token?

We do already, we just check if it is connected to the fee token for a different method, the `is_overlapping` method, that should return `true` if it would be theoretically possible to find a valid solution for the given orderbook, which require a connection to the fee token.

> This would allow us to keep the find_cycle logic private to our version of the bellman_ford search.

I actually prefer that the logic was moved out and into its own module (with, in one of the next PRs adds yet another method for determining paths) as it keeps the Bellman-Ford implementation just a Bellman-Ford implementation without any extra code or logic not related  to the algorithm. Additionally a predecessor vector (or mapping) is strictly more information than just a path along a negative cycle which is nice.

I eventually a nice abstraction would be a `Predecessors` type which has a predecessor vector or mapping as its internal implementation and provides methods for extracting paths from those.

Reply:
Ok. Just a nit that if we change the interface, I think we should lift predecessors out of the subtypes and make this method return a (Predecessors, Result) tuple. 
It seems that NegativeCycle is the only possible error, and thus this method is expected to always return its predecessors (regardless of error or not).

Reply:
Hmm, my issue with this is that the predecessors vectors are slightly different. In the `Ok` case they contain information to retrieve the shortest path to each connected node from the starting node, in an `Err` case they contain a path that may be part of a negative cycle and, therefore, does not necessarily lead back to the starting node.

I think the most "correct" thing to do type-wise is to have separate new-types for each of these predecessor vectors that can be operated on differently (for example, in the `Ok` case, there is no sense in having a `find_cycle` method since it will never exist)

Also I'm not a big fan of `(..., Result, ...)` tuple types since you lose the ability to use things like `?` as well as integration with other crates (like anyhow `Context` for example) and are in general a little less ergonomic to use.

I do agree that the search and path finding types are a little underwhelming and could use some work to better express the data and how it can be used.

Reply:
So the way I envisioned it can return a `Result<Path, NegativeCycle>` with something (loosely) like this:
```
type Predecessors = Vec<Option<NodeIndex>>;
struct Path(NodeIndex, Predecessors);
impl Path {
    fn path_to(&self, other: NodeIndex) -> Option<Vec<NodeIndex>>;
    fn connected_nodes(&self) -> impl Iterator<Item = NodeIndex>;
}
struct Cycle(NodeIndex, Predecessors);
impl Cycle {
    fn cycle(&self) -> Vec<NodeIndex>;
    fn connected_nodes(&self) -> impl Iterator<Item = NodeIndex>;
}
```